saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/registration/OpenSamlAssertingPartyMetadataRepository.java

```diff
@@ -0,0 +1,297 @@
+ * limitations under the License.
+ */
+
+package org.springframework.security.saml2.provider.service.registration;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.function.Consumer;
+
+import javax.annotation.Nonnull;
+
+import net.shibboleth.utilities.java.support.component.ComponentInitializationException;
+import net.shibboleth.utilities.java.support.resolver.CriteriaSet;
+import net.shibboleth.utilities.java.support.resolver.ResolverException;
+import org.opensaml.core.criterion.EntityIdCriterion;
+import org.opensaml.core.xml.config.XMLObjectProviderRegistrySupport;
+import org.opensaml.saml.criterion.EntityRoleCriterion;
+import org.opensaml.saml.metadata.resolver.MetadataResolver;
+import org.opensaml.saml.metadata.resolver.filter.impl.SignatureValidationFilter;
+import org.opensaml.saml.metadata.resolver.impl.ResourceBackedMetadataResolver;
+import org.opensaml.saml.metadata.resolver.index.impl.RoleMetadataIndex;
+import org.opensaml.saml.saml2.metadata.EntityDescriptor;
+import org.opensaml.saml.saml2.metadata.IDPSSODescriptor;
+import org.opensaml.security.credential.Credential;
+import org.opensaml.security.credential.impl.CollectionCredentialResolver;
+import org.opensaml.xmlsec.config.impl.DefaultSecurityConfigurationBootstrap;
+import org.opensaml.xmlsec.signature.support.SignatureTrustEngine;
+import org.opensaml.xmlsec.signature.support.impl.ExplicitKeySignatureTrustEngine;
+
+import org.springframework.core.io.DefaultResourceLoader;
+import org.springframework.core.io.Resource;
+import org.springframework.core.io.ResourceLoader;
+import org.springframework.lang.NonNull;
+import org.springframework.security.saml2.Saml2Exception;
+import org.springframework.security.saml2.core.OpenSamlInitializationService;
+import org.springframework.security.saml2.core.Saml2X509Credential;
+
+/**
+ * An implementation of {@link AssertingPartyMetadataRepository} that uses a
+ * {@link MetadataResolver} to retrieve expire-aware refreshing
+ * {@link AssertingPartyMetadata} instances.
+ *
+ * @author Josh Cummings
+ * @since 6.4
+ * @see AssertingPartyMetadataRepository
+ * @see RelyingPartyRegistrations
+ */
+public final class OpenSamlAssertingPartyMetadataRepository implements AssertingPartyMetadataRepository {
+
+	static {
+		OpenSamlInitializationService.initialize();
+	}
+
+	private final MetadataResolver metadataResolver;
+
+	public OpenSamlAssertingPartyMetadataRepository(MetadataResolver metadataResolver) {
+		this.metadataResolver = metadataResolver;
+	}
+
+	@Override
+	@NonNull
+	public Iterator<AssertingPartyMetadata> iterator() {
+		CriteriaSet all = new CriteriaSet(new EntityRoleCriterion(IDPSSODescriptor.DEFAULT_ELEMENT_NAME));
+		Iterator<EntityDescriptor> descriptors = resolve(all).iterator();
```

Comment:
Whether this works depends on how the resolver is set up. If there's no index for `EntityRoleCriterion` then it will return nothing. Further, some resolvers e.g. `ChainingMetadataResolver` will not return all possible matches.

It can work with `EvaluableEntityRoleEntityDescriptorCriterion`, but only if the resolver has `resolveViaPredicatesOnly` set to `true`. 

You also cannot rely on `IterableMetadataSource` because not all classes implement that. You can special-case some classes to iterate over their delegates, but again not all of those may be iterable.

Basically, there's no way to implement this without restricting the configuration of resolvers that can be used. Looking for `IterableMetadataSource` is probably the better way.

Reply:
e.g.
```java
Stream<EntityDescriptor> allEntities(MetadataResolver resolver) {
    if (resolver instanceof IterableMetadataSource iterable) {
        return StreamSupport.stream(iterable.spliterator(), false);
    } else if (resolver instanceof ChainingMetadataResolver chaining) {
        return chaining.getResolvers().stream().flatMap(this::allEntities);
    } else if (resolver instanceof CompositeMetadataResolver composite) {
        return composite.getResolvers().stream().flatMap(this::allEntities);
    } else {
        return Stream.empty();
    }
}
```